### DISCUSSION TITLE
`RearchConsumer` Listeners handling

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#54] [Q&A]
Scenario: I want to subscribe to a `ChangeNotifier` on `callonce` and unsubscribe on `dispose` (`effect`).

Solution I: Declare the listener as a class method. This way I need to declare a late final class field that the listener needs to call. So I need to initialize this field (with `callonce` and `use.call`). I don't like too much this solution. This way I can easily un/susbcribe the listener (to `GoRouter`).

Solution II: Create an internal listener using `use.value(() {})`. Seems weird to do that, but should work.

What approach will you use?

Also I want to know if it is feasible to have `WidgetHandle` available in all class methods, not just in `build()`

**Comment by GregoryConrad** [comment]
> Scenario: I want to subscribe to a `ChangeNotifier` on `callonce` and unsubscribe on `dispose` (`effect`).

Sounds about right. Why can't you just do that directly?

> Also I want to know if it is feasible to have `WidgetHandle` available in all class methods, not just in `build()`

You probably shouldn't have any other methods in the `RearchConsumer`, other than maybe a basic helper method (although this will make refactoring to `@rearchWidget` more difficult in the future if you chose to embrace that syntax when it is released).

**Comment by busslina** [comment]
> Sounds about right. Why can't you just do that directly?

Because I need to use in the listener an object obtained via capsule and I cannot pass this object to the listener as argument because the listener signature must be `void Function()`. And If I wrap the listener into a function that receives the object as argument, then I cannot unsubscribe because I will not have the same listener reference when I do `use.effect(...)`

**Reply by GregoryConrad** [reply]
> Because I need to use in the listener an object obtained via capsule and I cannot pass this object to the listener as argument because the listener signature must be void Function(). 

So can you pass in a new function as the listener that will capture the outside environment? I.e.:

```dart
final myData = use(myDataCapsule);
changeNotifier.listen(() {
  doSomethingWith(myData);
});
```

**Reply by GregoryConrad** [reply]
You can also safely call `use(myDataCapsule)` in a `use.callonce`, in case you were wondering.

**Comment by busslina** [comment]
> You probably shouldn't have any other methods in the RearchConsumer, other than maybe a basic helper method (although this will make refactoring to @rearchWidget more difficult in the future if you chose to embrace that syntax when it is released).

I think I can use inheritance (extending classes) defining the build method in the parent class and implementing abstract builder (Widget) methods in the leaf classes. With also other util abstract methods like `requestFocus()`.

I will dig into to find the best approach for me but also following your best practices.

No more questions for today.

Thanks ;)
